lib/handler/gzip.c

```diff
@@ -0,0 +1,184 @@
+#include <assert.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <zlib.h>
+#include "h2o.h"
+
+#define GZIP_ENCODING 16
+#define DEFAULT_WBITS 15
+#define DEFAULT_MEMLEVEL 8
+#define BUF_SIZE 8192
+
+typedef struct st_gzip_encoder_t {
+    h2o_ostream_t super;
+    z_stream zstream;
+    int started;
+    H2O_VECTOR(h2o_iovec_t) bufs;
+} gzip_encoder_t;
+
+static void *gzip_encoder_alloc(void *opaque, unsigned int items, unsigned int size)
+{
+    return h2o_mem_alloc(items * size);
+}
+
+static void gzip_encoder_free(void *opaque, void *address)
+{
+    free(address);
+}
+
+static void send_gzip(h2o_ostream_t *_self, h2o_req_t *req, h2o_iovec_t *inbufs, size_t inbufcnt, int is_final)
+{
+    gzip_encoder_t *self = (void *)_self;
+    
+    size_t in_total, i, outbuf_count = 1;
+    
+    /* calc in data total size */
+    in_total = 0;
+    for (i = 0; i != inbufcnt; ++i)
+        in_total += inbufs[i].len;
+
+    if (self->started == 0) {
+        /* Initialization */
+        int ret = deflateInit2(&self->zstream, Z_DEFAULT_COMPRESSION, Z_DEFLATED,
+                               DEFAULT_WBITS + GZIP_ENCODING, DEFAULT_MEMLEVEL, Z_DEFAULT_STRATEGY);
+
+        assert(ret == Z_OK);
+
+        h2o_vector_reserve(&req->pool, (h2o_vector_t*)&self->bufs, sizeof(h2o_iovec_t),
+                           self->bufs.capacity + 1);
+        self->bufs.entries[0].base = h2o_mem_alloc_pool(&req->pool, BUF_SIZE);
```

Comment:
Well... I'm not sure if I was misunderstanding but this part only runs at the first call of `send_gzip` of one output stream. I check started flag to do so. And if this block is not enough, it will allocate more blocks as output buffer as code below.